From 50d6d8ecced8c93bfefe7f60806fcf669f0193d5 Mon Sep 17 00:00:00 2001
From: Vladimir Oltean <olteanv@gmail.com>
Date: Thu, 26 Dec 2019 04:20:27 +0200
Subject: [PATCH buildroot 1/4] net: dsa: sja1105: Request SPI timestamp for
 correct word

With the change to use bits_per_word=32, the sja1105 driver no longer
uses the correct API for PTP system timestamping (something that is
required for proper cross-timestamping with the value of the system
clock at the time of the SPI transactions).

This fixes a variety of random PTP bugs, most notably clock stepping
with phc2sys if SYSOFF_EXTENDED is being used.

Fixes: 23ec4bb3ab87 ("net: dsa: sja1105: Increase SPI word size to 32 bits")
Signed-off-by: Vladimir Oltean <olteanv@gmail.com>
---
 drivers/net/dsa/sja1105/sja1105_spi.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/dsa/sja1105/sja1105_spi.c b/drivers/net/dsa/sja1105/sja1105_spi.c
index 29b127f3bf9c..4c599109f78d 100644
--- a/drivers/net/dsa/sja1105/sja1105_spi.c
+++ b/drivers/net/dsa/sja1105/sja1105_spi.c
@@ -119,8 +119,8 @@ static int sja1105_xfer(const struct sja1105_private *priv,
 			ptp_sts_xfer = hdr_xfer;
 		else
 			ptp_sts_xfer = chunk_xfer;
-		ptp_sts_xfer->ptp_sts_word_pre = ptp_sts_xfer->len - 1;
-		ptp_sts_xfer->ptp_sts_word_post = ptp_sts_xfer->len - 1;
+		ptp_sts_xfer->ptp_sts_word_pre = (ptp_sts_xfer->len - 1) / 4;
+		ptp_sts_xfer->ptp_sts_word_post = (ptp_sts_xfer->len - 1) / 4;
 		ptp_sts_xfer->ptp_sts = ptp_sts;
 
 		/* Calculate next chunk */
-- 
2.17.1

